### PR TITLE
Allow user defined coverage excludes

### PR DESCRIPTION
**Pull request by kjerstadius**
Addresses #3287. 

This is my first pull request for this project, so I'm aware that I may have mistakenly overlooked certain aspects. Please let me know if that's the case.

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -687,14 +687,20 @@ def generate_run_target(self, target, outfile):
...
         elem.add_item('COMMAND', self.environment.get_build_command() +
                       ['--internal', 'coverage'] +
                       outputs +
                       [self.environment.get_source_dir(),
-                       os.path.join(self.environment.get_source_dir(),
-                                    self.build.get_subproject_dir()),
                        self.environment.get_build_dir(),
-                       self.environment.get_log_dir()])
+                       self.environment.get_log_dir()] +
+                       coverage_excludes)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__ [E127] continuation line over-indented for visual indent

<sub>posted by Sider</sub>

File: mesonbuild/scripts/coverage.py [hunk 2]
```diff
@@ -85,11 +88,11 @@ def coverage(outputs, source_root, subproject_root, build_root, log_dir):
...
                                    '--output-file', covinfo])
-            # Remove all directories inside subproject dir
+            # Remove all files/directories defined as excludes
             subprocess.check_call([lcov_exe,
-                                   '--remove', covinfo,
-                                   os.path.join(subproject_root, '*'),
-                                   '--rc', 'lcov_branch_coverage=1',
+                                   '--remove', covinfo] +
+                                   lcov_excludes +
+                                   ['--rc', 'lcov_branch_coverage=1',
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/coverage.py]: __[Flake8]__ [E127] continuation line over-indented for visual indent

<sub>posted by Sider</sub>

File: mesonbuild/scripts/coverage.py [hunk 3]
```diff
@@ -85,11 +88,11 @@ def coverage(outputs, source_root, subproject_root, build_root, log_dir):
...
                                    '--rc', 'lcov_branch_coverage=1',
                                    '--output-file', covinfo])
-            # Remove all directories inside subproject dir
+            # Remove all files/directories defined as excludes
             subprocess.check_call([lcov_exe,
-                                   '--remove', covinfo,
-                                   os.path.join(subproject_root, '*'),
-                                   '--rc', 'lcov_branch_coverage=1',
+                                   '--remove', covinfo] +
+                                   lcov_excludes +
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/coverage.py]: __[Flake8]__ [E127] continuation line over-indented for visual indent

<sub>posted by Sider</sub>

**Comment by jpakkane** [comment]
We try to avoid adding new options wheneve possible because otherwise you end up with a gazillion of them which is bad UX. How is this feature used? Are the things you'd want to exclude from gcovr static (i.e. the `tests` subdirectory and nothing else) or are they dynamic, meaning you want to change which directories are ignored frequently?

**Comment by kjerstadius** [comment]
I can understand that and I agree that thoughtlessly adding new options all the time isn't where you'd want to go.

In my specific case, the need is a combination of a `test`/`tests` folder and a few stray files elsewhere. Maybe my case is atypical, but being able to freely define excludes seems quite useful to me. To answer your question though, the files don't really change from one build to the next, though there's of course a possibility that new files are added to the project that we'd want to exclude. 

Anyway, is there any other way we could define excludes that isn't a built-in option? I was thinking that perhaps it would be possible to have an option that the user would define in their meson_options.txt if they have the need, but then I thought that maybe it goes against the idea of what project options are supposed to be.

**Comment by kjerstadius** [comment]
I guess another option is to introduce a coverage_exclude() function for use in the build files. I'm open to suggestions if you think making this functionality available to the user is a good idea :)

**Comment by jpakkane** [comment]
I filed [an upstream issue on gcovr](https://github.com/gcovr/gcovr/issues/293) about this since it seems generally useful. Let's see what they have to say about this first.

**Comment by jpakkane** [comment]
Does the new feature described by gcovr upstream in the linked bug fix things for you?

**Comment by kjerstadius** [comment]
I think it would! Though I'm a little confused as to exactly what is meant by 

> The list of exclusions is not currently provided in the HTML report, and I don't know whether that would be helpful in the general case.

I'll get gcovr from git and see if I can test it locally to confirm the behaviour. Thanks!

**Comment by kjerstadius** [comment]
After reading the comment I thought confusing above a few more times I realised I was just being a little slow...

Anyway, so I tried to test the master version of gcovr, but unfortunately I get no numbers at all. I.e. I use the same setup I normally do, just replacing the gcovr package with a local version of gcovr from git (using pip -e), but when I run `ninja coverage` no files get picked up and subsequently I get 0% coverage. Until I can get that resolved I won't be able to verify that the config file solution works.

**Comment by jpakkane** [comment]
Since our MR tracker is getting quite bloated, I'm going to close this. Feel free to reopen if the fix does not work for you. Thanks.

**Comment by perror** [comment]
Is there anything new about the possibility to exclude the tests themselves from the code coverage process ?

**Pull request closed without merging**